File: LeetCode/DP/Easy/Best Time to Buy and Sell Stock.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
"""
1 <= prices.length <= 10^5
0 <= prices[i] <= 10^4
"""
class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        leng = len(prices)
        dp = [0] * leng
        for i in range(1, leng):
            dp[i] = max(dp[i-1] + prices[i] - prices[i-1], 0)  # 직접 살펴봐야 이해하기 쉬움.
        return max(dp)
        # min_price = prices[0]
        # max_profit = 0
        # for i in range(1, leng):
        #     min_price = min(min_price, prices[i])
        #     max_profit = max(max_profit, prices[i] - min_price)
        # return max_profit

inputdatas = [
    [7,1,5,3,6,4],
    [7,6,4,3,1]
]

import inspect
sol = Solution()
functions = inspect.getmembers(sol, predicate=inspect.ismethod)
for t in inputdatas:
    print(functions[0][1](t))

"""
LeetCode Easy. dp로 먼저 접근했다. DP가 항상 어렵게 느껴졌고, 
시간을 엄청 많이 쓴건 아니지만 Easy임에도 약간 어렵게 느껴졌다.
아래 풀이처럼 최소구매가격, 최대수익을 갱신하는 방식으로 풀이하면 더 직관적이다.
min max 대신 if문을 사용하면 시간복잡도가 약간 줄어드는 것 같다.
"""

File: LeetCode/DP/Easy/Best Time to Buy and Sell Stock.js
// https://leetcode.com/problems/count-number-of-nice-subarrays/

/*
constraints:
*/


/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    const leng = prices.length;
    const dp = Array(leng).fill(0)
    for (let i = 1; i < leng; i++) {
        dp[i] = Math.max(dp[i-1] + prices[i] - prices[i-1], 0);
    }
    return Math.max(...dp);
};


const inputDatas = [
    {data: [[7,1,5,3,6,4]], answer: 5},
    {data: [[7,6,4,3,1]], answer: 0},
];

/*
LeetCode Easy.
제출 9.8M, 정답률 54.2%
dp[i]는 i번째 날까지 얻을 수 있는 최대 이익.
Math.max의 첫 인자는 i-1 최대이익에 i번째 날 이익을 더한 값,
두번째 인자는 0으로 둬서 마이너스 대비(안사는 경우)
i-1만 체크하기 때문에 이전에 한번 사고 판 수익을 참조해서 두번 이상 사고파는 경우는 생기지 않음.
파이썬 풀이를 그대로 사용했다. dp연습중.
*/

for (let i = 0; i < inputDatas.length; i++) {
    const {data, answer} = inputDatas[i];
    const res = maxProfit(...data);
    if (JSON.stringify(res) === JSON.stringify(answer)) {
        console.log("pass");
    } else {
        let summary = "fail";
        for (const [label, content] of [["expected:", answer], ["got:", res]]) {
            summary += `\n  ${label}\n`;
            summary += `    ${content}`;
            summary = summary.trimEnd();
        }
        console.log(summary);
    }
}